metersink/output_odoo.py

```diff
@@ -129,157 +236,323 @@ def get_odoo_user_id(odoo):
+            )
+        else:
+            # If there is no line already for the ressource, create it.
+
+            time_calc = calculate_cloud_time(
+                data_dict["traits"]["created_at"]
+            )
+
+            line_dict = {
+                "product_id": 36,
+                "name": display_name,
+                "product_uom_qty": time_calc,
+                "order_id": sale_orders["id"],
+            }
+            line_id = odoo_create(odoo, "sale.order.line", [line_dict])
+            LOG.debug("%s", line_id)
+
+
+def create_sale_order(odoo, customer, tag_list):
+    """creates a new SO for customer with tags"""
+    record_list = [
+                {
+                    "partner_id": customer['id'],
+                    "category_ids": tag_list
+                }
+            ]
+    new_id = odoo_create(odoo, "sale.order", record_list)
+    return new_id
+
+
+def create_sale_order_line(odoo, order_id, product_id, display_name, product_uom_qty):
+    """creates a new so line and returns its id"""
+    record_list = [
+        {
+            "order_id": order_id,
+            "product_id": product_id,
+            "display_name": display_name,
+            "product_uom_qty": product_uom_qty,
+        }
+    ]
+    new_id = odoo_create(odoo, "Sale.order.line", record_list)
+    return new_id
+
+
+def get_sale_order_id(odoo, tag_list):
     """
-    looks for already existing entries.
+    Returns the id of a Sale_order to further work on if the Customer is known.
+    Else it gives nothing back.
     """
-    # print(line)
-    if not filter_dict:
-        filter_dict = {}
+    # project_tag = f"project={project_id}"
+
+    [customer] = odoo_get_customer_from_project(odoo, tag_list)
+    if customer:
+        print(pformat(customer))
+        print(customer["sale_order_ids"])
+
+        if LOG.isEnabledFor(logging.DEBUG):
+            show_sale_order_fields(odoo)
+        filter_list = [
+            #"|",
+            ["customer_id" , "=", customer["id"]],
+            #["category_ids", "=", project_tag]
+        ]
+        projection_dict = get_projection_dict()
+        [sale_order_id] = odoo_get(
+            odoo,
+            "sale.order",
+            mode="ids",
+            filter_list=filter_list,
+            projection_dict=projection_dict,
+        )
 
-    filter_dict["date"] = str(datetime.date(line["start"]))[:10]
-    filter_dict["name"] = "TimeWarrior import " + get_hash(line["start"])
-    filter_dict["employee_id"] = get_employee_id(odoo)
+        if not sale_order_id:
+            # create new sale order
+            sale_order_id = create_sale_order(odoo, customer, tag_list)
+        return sale_order_id
+    return None
 
-    filter_list = []
-    for key, value in filter_dict.items():
-        filter_list.append([key, "=", value])
 
-    project_id = get_project_id(odoo, line["tags"])
+def get_sale_order_line():
+    """to be implemented"""
+    pass
```

Comment:
This function has no parameter and the docstring says "to be implemented", so it can be dropped altogether?

Reply:
It is implemented now.